src/utils/rootkeypackage_utils/src/rootkeypackage_parse.c

```diff
@@ -0,0 +1,827 @@
+    {
+        Log_Error("ERC %d parsing 'protected' property.", result.ResultCode);
+    }
+
+    return result;
+}
+
+/**
+ * @brief Parses the signatures properties in accordance with rootkeypackage.schema.json
+ *
+ * @param rootObj The root JSON object.
+ * @param[out] outPackage The root key package object to write parsed signatures data.
+ *
+ * @return The result.
+ */
+ADUC_Result RootKeyPackage_ParseSignatures(JSON_Object* rootObj, ADUC_RootKeyPackage* outPackage)
+{
+    ADUC_Result result = { .ResultCode = ADUC_GeneralResult_Failure, .ExtendedResultCode = 0 };
+    VECTOR_HANDLE hashes = NULL;
+
+    JSON_Array* signaturesArray = json_object_get_array(rootObj, "signatures");
+    if (signaturesArray == NULL)
+    {
+        result.ExtendedResultCode = ADUC_ERC_UTILITIES_ROOTKEYPKG_PARSE_MISSING_REQUIRED_PROPERTY_SIGNATURES;
+        goto done;
+    }
+
+    size_t cnt = json_array_get_count(signaturesArray);
+    if (cnt == 0)
+    {
+        result.ExtendedResultCode = ADUC_ERC_UTILITIES_ROOTKEYPKG_PARSE_SIGNATURES_EMPTY;
+        goto done;
+    }
+
+    hashes = VECTOR_create(sizeof(ADUC_RootKeyPackage_Hash));
+    if (hashes == NULL)
+    {
+        result.ExtendedResultCode = ADUC_ERC_NOMEM;
+        goto done;
+    }
+
+    for (size_t i = 0; i < cnt; ++i)
+    {
+        ADUC_RootKeyPackage_Hash hashElement = { .alg = ADUC_RootKeyShaAlgorithm_INVALID, .hash = NULL };
+        ADUC_RootKeyShaAlgorithm tmpAlg = ADUC_RootKeyShaAlgorithm_INVALID;
+        CONSTBUFFER_HANDLE hashBuf = NULL;
+
+        JSON_Object* hashJsonArrayElementObj = json_array_get_object(signaturesArray, i);
+        if (hashJsonArrayElementObj == NULL)
+        {
+            result.ExtendedResultCode = ADUC_ERC_UTILITIES_ROOTKEYPKG_PARSE_GETOBJ_SIGNATURES_ELEMENT;
+            goto done;
+        }
+
+        result = RootKeyPackage_ParseShaHashAlg(hashJsonArrayElementObj, &tmpAlg);
+        if (IsAducResultCodeFailure(result.ResultCode))
+        {
+            goto done;
+        }
+
+        result = RootKeyPackage_ParseShaHashValue(hashJsonArrayElementObj, "sig", &hashBuf);
+        if (IsAducResultCodeFailure(result.ResultCode))
+        {
+            goto done;
+        }
+
+        hashElement.alg = tmpAlg;
+        hashElement.hash = hashBuf;
+
+        if (VECTOR_push_back(hashes, &hashElement, 1) != 0)
+        {
+            // can't add to vector, so free it
+            CONSTBUFFER_DecRef(hashElement.hash);
+            hashElement.hash = NULL;
+
+            result.ExtendedResultCode = ADUC_ERC_NOMEM;
+            goto done;
+        }
+        memset(&hashElement, 0, sizeof(hashElement));
+    }
+
+    outPackage->signatures = hashes;
+    hashes = NULL;
+
+    result.ResultCode = ADUC_GeneralResult_Success;
+done:
+
+    if (hashes != NULL)
+    {
+        size_t cnt = VECTOR_size(hashes);
+        for (size_t i = 0; i < cnt; ++i)
+        {
+            ADUC_RootKeyPackage_Hash* node = (ADUC_RootKeyPackage_Hash*)VECTOR_element(hashes, i);
```

Comment:
do we need to free node?

Reply:
yes, need to call ADUC_RootKeyPackage_Hash_DeInit() instead of the stuff on line 815.  I thought I had replaced all of these with said function, but missed this one.  Fixed.